fix mms-agent recipe to not clobber the whole installation dir:

mattchukabam opened the pull request:
The existing installation process blows away its installation dir, which is not
polite if it's /usr/local/share (which is the default); plenty of other
software installs stuff there.

Changes:
- mms_agent.install_dir now correctly points to the actual install dir, and not
  its parent
- use chef's file cache instead of /tmp
- improved handling of installation dir
- both the installation and use of munin is now configurable via attribute

mattchukabam committed:
fix mms-agent recipe to not clobber the whole installation dir:

The existing installation process blows away its installation dir, which is not
polite if it's /usr/local/share (which is the default); plenty of other
software installs stuff there.

Changes:
- mms_agent.install_dir now correctly points to the actual install dir, and not
  its parent
- use chef's file cache instead of /tmp
- improved handling of installation dir
- both the installation and use of munin is now configurable via attribute

josephholsten commented:
@mattchukabam looks nice. Any chance you could write tests for test-kitchen or chef-spec? I'd like to make sure we don't break this in future.

jamesonjlee commented:
:+1: to above idea

mattchukabam committed:
initial chefspecs for mms-agent recipe (ubuntu only)

mattchukabam commented:
Went with chefspec (to learn). Feedback appreciated.

jamesonjlee commented:
looks right, i'll give it a run and then merge.

jamesonjlee commented:
comments from running this locally:
- the 3rd test runs real commands on the local client, (known issue with chefspec), probably should take it out
- I don't think mms-agent needs to do the `rm -rf`?  it uses `zip -o` ?
- (optional)  change `node['mongodb']['mms_agent']['logdir']` from `node['mongodb']['logpath']` to something like `File.dirname(node['mongodb']['config']['logpath']/agent`
- (optional) add test to check that the runit service was added (but has not ran)
- (optional) isn't remote_file suppose to verify if it needs to redownload the file between runs?
- (optional) rearrange includes/packages to bring static/simple resources up top, and complex ones down.

jamesonjlee commented:
also, thanks for taking time to write this PR :)

mattchukabam committed:
improvements:

- use the new `node['mongodb']['config']['logpath']` attribute
- marked as pending test that actually modifies the host machine

mattchukabam committed:
added new attribute , in case the agent download fails

mattchukabam commented:
Took me a while to get back to this:
1. A few days ago the mms-agent download link broken (would hang forever whenever we tried to download). This broke all of our mms-agent enabled instances, so we added the new `ignore_failure_on_install` attribute. I think this is a reasonable addition, especially if the default is changed from `ignore_failure_on_install = true` to `false` so that there's no difference at all in behavior.
2. Couldn't figure out how to write an runit service test in chefspec.
3. `remote_file` does verify whether it needs to download, but the checksum on the mongo side isn't constant (see the referenced issue in the comments).

josephholsten commented:
um, I'm not comfortable with an `ignore_failure_on_install` option. Either they should want monitoring on, or not.

As for the remote_file checksum, I wouldn't worry about it with the new Chef 11 [file provider improvements](https://gist.github.com/danielsdeleo/5732547#file-ft-announce-md). Assuming the file server sets either an ETag or  Last-Modified header, chef will cache it reasonably now.

Thanks for making an attempt with chef spec, I'll dig in soon to see if there's anything else I can add!

mattchukabam commented:
I agree with "they should want monitoring on, or not"; I want monitoring on, but if they break something upstream I also don't want it to break my entire chef-run. That being said, I'm still okay with taking it out; it is a bad option.

josephholsten commented:
@mattchukabam also relevant to your interests: https://github.com/edelight/chef-mongodb/pull/185

For situations like this (_cough_ oracle-java-and-mysql _cough_), we tend to stick assets on a local file server so upstream can't break stuff.

mattchukabam committed:
Revert "added new attribute , in case the agent download fails"

This reverts commit 057be15cfeab76dfc4b7c2a8077fe52cbc363e78.

mattchukabam commented:
Fair enough, reverted. It's in the history so it should be easy to add it in again if the need arises. Oh, and the URL's already updated in this PR.
